fsdev/db.cpp

```diff
@@ -0,0 +1,238 @@
+	{
+		if (i > 0)
+		{
+			query_columns += ", ";
+			parameter_columns += ", ";
+		}
+		query_columns += track_column_names[i];
+		parameter_columns += "?";
+	}
+	query_columns += ") ";
+	parameter_columns += ") ";
+}
+
+bool TruncateTables(sql::Connection * connection, Logger & _log_)
+{
+	bool rv = false;
+	sql::Statement * stmt = connection->createStatement();
+	try
+	{
+		if (stmt)
+		{
+			stmt->execute("truncate paths;");
+			stmt->execute("truncate tracks;");
+			delete stmt;
+			rv = true;
+		}
+	}
+	catch (sql::SQLException & e)
+	{
+		stringstream ss;
+		ss << "# ERR: SQLException in " << __FILE__;
+		ss << "(" << __FUNCTION__ << ") on line " << __LINE__ << endl;
+		ss << "# ERR: " << e.what();
+		ss << " (MySQL error code: " << e.getErrorCode();
+		ss << ", SQLState: " << e.getSQLState() << " )" << endl;
+		throw LOG2(_log_, ss.str(), LogLevel::FATAL);
+	}
+	return rv;
+}
+
+string SanitizeString(string & s)
+{
+	string rv;
+	rv.reserve(s.size() * 2);
+	for (size_t i = 0; i < s.size(); i++)
+	{
+		switch (s[i])
+		{
+			case '\'':
+			rv += "''";
+			break;
+
+			default:
+			rv += s[i];
+			break;
+		}
+	}
+	return rv;
+}
+
+bool InsertDirectory(sql::Connection * connection, DIRENT * d, string & nspace, Logger & _log_)
+{
+	bool rv = false;
+
+	sql::Statement * stmt = connection->createStatement();
+	try
+	{
+		if (stmt)
+		{
+			string s;
+			s = "insert into paths (me, up, name, namespace) values(";
+			s += to_string(d->me) + ", ";
+			s += to_string(d->up) + ", ";
+			s += "\'" + SanitizeString(d->name) + "\', ";
+			s += "\'" + SanitizeString(nspace) + "\')";
+			s += ";";
+			stmt->execute(s.c_str());
+			rv = true;
+		}
+	}
+	catch (sql::SQLException & e)
+	{
+		stringstream ss;
+		ss << "# ERR: SQLException in " << __FILE__;
+		ss << "(" << __FUNCTION__ << ") on line " << __LINE__ << endl;
+		ss << "# ERR: " << e.what();
+		ss << " (MySQL error code: " << e.getErrorCode();
+		ss << ", SQLState: " << e.getSQLState() << " )" << endl;
+		throw LOG2(_log_, ss.str(), LogLevel::FATAL);
+	}
+	return rv;
+}
+
+bool InsertTrack(sql::Connection * connection, DIRENT * d, string & nspace, Logger & _log_)
+{
+	bool rv = false;
+	FILE * p;
+
+	if (access(d->fullpath.c_str(), R_OK) < 0)
+	{
+		cerr << "cannot access: " + d->fullpath << endl;
+	}
+
+	string cmdline = string("ffprobe -loglevel quiet -show_entries stream_tags:format_tags -show_entries format=duration -sexagesimal \"") + d->fullpath + string("\"");
+
+	if ((p = popen(cmdline.c_str(), "r")) == nullptr)
+	{
+		perror("pipe to ffprobe failed");
+		return rv;
+	}
+
+	const int bsize = 1024;
+	char buffer[bsize] = { 0 };
+
+	Track track;
+
+	while (fgets(buffer, bsize, p) != nullptr)
+	{
+		string b(buffer);
+		track.SetTag(b);
+		memset(buffer, 0, bsize);
+	}
+
+	pclose(p);
+
+	try
+	{
+		size_t n = sizeof(track_column_names) / sizeof(string);
+		sql::PreparedStatement * stmt = nullptr;	
+		string s = "insert into tracks " + query_columns + parameter_columns + ";";
```

Comment:
This logic could also be moved outside of the try/catch.
I prefer only running the code that would cause the caught exception in the try block.